---
title: "Getting_Sequences_to_Analyze"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Getting_Sequences_to_Analyze}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(RNAreachR)
```

# Getting started with RNAreachR
  RNAreachR is a package focused on comparing sets of RNA sequence
It is expected that useres will have some set of sequences that they are expecially interested in.
These are usually defined by some sort of RNA sequencing experiment.
For example, significantly differentially expressed genes.

  It is most useful to define interesting transcripts as transcripts that are highly expressed in your experimental system.
These transcripts can then be used to extract interesting sequence.

  However, often an experiment reveals an interesting set of genes and returning to the expression data may not be possible.
This vignette is designed to help users go from a set of genes they are interested in to sequence.

This aspect of RNAreachR is currently only compatible with Gencode human and mouse data.

##Reading in Gencode gffs and filtering them
  In this vignette we will work with some example mouse data from an RNA sequencing experiment.
First, we will need the mouse annotations from gencode.
The most recent version can be downloaded from here: \url{https://www.gencodegenes.org/mouse/}
This gff file contains the positional information for every gene, transcript, UTR and exon in the mouse genome.
Because we aren't usually interested in every transcript, will first filter this file using \code{filter_Tx}

```{r, get gffs and filter}
mm_filtered_TxDb <- filter_Tx("mydata/Gencodedat/gencode.vM20.annotation.gff3.gz")
mm_filtered_TxDb
```
  \code{filter_Tx} creates a TxDb object. 
These objects create gene models that make accessing transcripts associated with genes very simple.
It further models coding regions, UTRs and all exons.
See \url{https://www.rdocumentation.org/packages/GenomicFeatures/versions/1.24.4/topics/TxDb-class} for more information about TxDb objects.
  \code{mm_filtered_TxDb} contains ~68,000 transcripts from the mouse genome. Using \code{filter_Tx} we have excluded all 
low confidence transcripts. We recommend this as many transcripts are annotated with very little evidence that they truely exist.
\code{filter_Tx} can also filter out non-coding transcripts. However, in this vignette we are still interested in non-coding RNAs.

##Converting genes to transcripts in a meaningful way
  In many experiments, a group of genes may become interesting for a variety of reasons.
The \code{case_genes} in this vignette are interesting because their localization in neuronal cells relies on FMRP expression.
These genes met the many many required thresholds from a previous analysis to be considered interesting.
\code{ctrl_genes} contains every other "uninteresting" gene expressed in the neuronal cells.
These gene sets are very different lengths.
  Because many different transcripts or isoforms are present within a single gene, we must determine which transcript we want to retrieve sequence from to best represent the genes.
It is best to use expression data to determine the most expressed transcript for each gene.
If that is not an option, we can select transcripts based on their lengths.
Often, the longest transcript or the most median length transcript is selected.
  \code{make_longest_df} and \code{make_median_df} use the filtered TxDb object to create dataframe that relate Ensembl gene IDs with the longest or most median length transcript. Further, this dataframe also contains the transcript with the longest coding region (CSD) or longest 5' or 3'UTR. 
  The dataframe generated by \code{make_longest_df} can be then be used to translate a gene list into a transcript list where each transcript has the longest feature for that gene. \code{gene2Tx} facilitates this translation and which sequence feature length is considered (whole transcript, CDS, 5'UTR or 3'UTR).
  
```{r, gene list to tx list}
#read in example gene lists
case_genes <- read.table("mydata/example_case_genes.txt") %>% pull(., x) %>% as.character()
ctrl_genes <- read.table("mydata/example_ctrl_genes.txt") %>% pull(., x) %>% as.character()

#convert to transcript lists
longest_mm <- make_longest_df(mm_filtered_TxDb)
case_longest_tx <- gene2Tx(longest_mm, case_genes, "UTR3")
ctrl_longest_tx <- gene2Tx(longest_mm, ctrl_genes, "UTR3")

#Alternatively, median length transcripts could be used
median_mm <- make_median_df(mm_filtered_TxDb)
case_median_tx <- gene2Tx(median_mm, case_genes, "UTR3")
ctrl_median_tx <- gene2Tx(median_mm, ctrl_genes, "UTR3")
```
The resulting transcript list may become shorter than the original gene list. 
This is because some genes only contain non-coding transcripts which lack CDS, 5'UTR and 3'UTR sequence.
##Exporting sequences of interesting transcripts
  Now we have both the filtered genome annotation TxDb object and transcript lists. We can extract the sequences of the transcripts we find interesting and store them in either fasta files, gff3 files or both.
```{r, tx list to sequence}
getTxOut(mm_filtered_TxDb, case_longest_tx, "UTR3", "mydata/example_case", "both")
getTxOut(mm_filtered_TxDb, ctrl_longest_tx, "UTR3", "mydata/example_ctrl", "both")
```
  With the \code{getTxOut} output files generated, the true function of RNAreachR can be used.
These functions which rely on case and control fastas or case and control gff3 files will be detailed in a different vignette.



